src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -280,78 +316,84 @@ class UserService(protected val userInfo: UserInfo, dataSource: DataSource, prot
+
+          (users.collect { case (email, None) => email }, subGroups.collect { case (email, None) => email }) match {
+            // success case, all users and groups found
+            case (Seq(), Seq()) =>
+              updateGroupMembersInternal(group, users.map(_._2.get), subGroups.map(_._2.get), adding, txn) map {
+                _ match {
+                  case None => RequestComplete(StatusCodes.OK)
+                  case Some(error) => RequestComplete(error)
                 }
               }
-            }
+
+            // failure cases, some users and/or groups not found
+            case (Seq(), missingGroups) => Future.successful(RequestComplete(ErrorReport(StatusCodes.NotFound, s"Some groups not found: ${missingGroups.mkString(", ")}")))
+            case (missingUsers, Seq()) => Future.successful(RequestComplete(ErrorReport(StatusCodes.NotFound, s"Some users not found: ${missingUsers.mkString(", ")}")))
+            case (missingUsers, missingGroups) => Future.successful(RequestComplete(ErrorReport(StatusCodes.NotFound, s"Some users not found: ${missingUsers.mkString(", ")}. Some groups not found: ${missingGroups.mkString(", ")}")))
           }
+        }
+      }
+    }
+  }
 
-          list.map { tries =>
-            val exceptions = tries.collect {
-              case Failure(t) => t
-            }
-
-            val successfulUsers = tries.collect {
-              case Success(Left(member)) => RawlsUser.toRef(member)
-            }.toSet
-            val successfulGroups = tries.collect {
-              case Success(Right(member)) => RawlsGroup.toRef(member)
-            }.toSet
-
-            val group = containerDAO.authDAO.loadGroup(groupRef, txn).getOrElse(throw new RawlsException("Unable to load group"))
-            adding match {
-              case true => containerDAO.authDAO.saveGroup(group.copy(users = (group.users ++ successfulUsers),
-                subGroups = (group.subGroups ++ successfulGroups)), txn)
-              case false => containerDAO.authDAO.saveGroup(group.copy(users = (group.users -- successfulUsers),
-                subGroups = (group.subGroups -- successfulGroups)), txn)
-            }
-            if (exceptions.isEmpty)
-              RequestComplete(StatusCodes.OK)
-            else
-              RequestComplete(ErrorReport(StatusCodes.BadRequest, "Unable to update the following member(s)", exceptions.map(ErrorReport(_))))
+  //in the event of a failure updating the user in the google group, the user will not be updated in the RawlsGroup
+  private def updateGroupMembersInternal(group: RawlsGroup, users: Seq[RawlsUser], subGroups: Seq[RawlsGroup], adding: Boolean, txn: RawlsTransaction): Future[Option[ErrorReport]] = {
+    val updateMap = users.map { user =>
+      val updateTry = adding match {
+        case true => toFutureTry(gcsDAO.addMemberToGoogleGroup(group, Left(user)))
+        case false => toFutureTry(gcsDAO.removeMemberFromGoogleGroup(group, Left(user)))
```

Comment:
this is prior code so I am not going to fiddle with it. I don't really like it either but have not really thought of something better. I don't think a 2 value enum is really any better.